RFS/src/main/java/com/rfs/common/LuceneDocumentsReader.java

```diff
@@ -64,51 +66,55 @@ public static Function<Path, LuceneDocumentsReader> getFactory(boolean softDelet
-                            .runOn(luceneReaderScheduler) // Specify thread to use on read calls on, disable prefetch
-                            .concatMapDelayError( // Delay errors to attempt reading all documents before exiting
-                                docIdx -> Mono.justOrEmpty((liveDocs == null || liveDocs.get(docIdx)) ? // Filter for live docs
-                                    getDocument(segmentReader, docIdx, true) : // Get document, returns null to skip malformed docs
-                                    null));
-                    }, segmentReader -> {
-                        // NO-OP, closed by top level reader.close()
-                    })
-                );
-        }, reader -> {
-            try {
-                reader.close();
-            } catch (IOException e) {
-                log.atError().setMessage("Failed to close DirectoryReader").setCause(e).log();
-                throw Lombok.sneakyThrow(e);
-            }
-            // Close scheduler
-            luceneReaderScheduler.dispose();
+        return Flux.using(
+            () -> wrapReader(DirectoryReader.open(FSDirectory.open(indexDirectoryPath)), softDeletesPossible, softDeletesField),
+            this::readDocsByLeavesInParallel,
+            reader -> {
+                try {
+                    reader.close();
+                } catch (IOException e) {
+                    log.atError().setMessage("Failed to close DirectoryReader").setCause(e).log();
+                    throw Lombok.sneakyThrow(e);
+                }
         });
     }
 
+    Publisher<Document> readDocsByLeavesInParallel(DirectoryReader reader) {
+        var segmentsToReadAtOnce = 5; // Arbitrary value
+        var maxDocumentsToReadAtOnce = 100; // Arbitrary value
+        log.atInfo().setMessage("{} documents in {} leaves found in the current Lucene index")
+            .addArgument(reader::maxDoc)
+            .addArgument(reader.leaves()::size)
+            .log();
+    
+        // Create shared scheduler for i/o bound document reading
+        var sharedSegmentReaderScheduler = Schedulers.newBoundedElastic(maxDocumentsToReadAtOnce, Integer.MAX_VALUE, "sharedSegmentReader");
+
+        return Flux.fromIterable(reader.leaves())
+            .parallel(segmentsToReadAtOnce)
+            .concatMapDelayError(leaf -> readDocsFromSegments(leaf, sharedSegmentReaderScheduler));
+    }
+
+    Publisher<Document> readDocsFromSegments(LeafReaderContext leafReaderContext, Scheduler scheduler) {
+        var documentsToReadAtOnceForEachSegment = 50; // Arbitrary value
+        var segmentReader = leafReaderContext.reader();
+        var liveDocs = segmentReader.getLiveDocs();
+
+        return Flux.range(0, segmentReader.maxDoc())
+            .parallel(documentsToReadAtOnceForEachSegment)
+            .runOn(scheduler) // Specify thread to use on read calls on, disable prefetch
+            .filter(docIdx -> (liveDocs == null || liveDocs.get(docIdx)))
```

Comment:
Moved this check away from the concatMap to ensure we are only processing the expected docs